### DISCUSSION TITLE
Agent Uninstalled - Log & Web Notification

### DISCUSSION THREAD

**Discussion by bayden10** [Ylianst/MeshCentral#6617] [Ideas]
Hello, 

In some cases AD or Group Policies are not available which leaves some endpoint users the ability to uninstall the Agent. Currently if an agent is uninstalled a response of "Agent disconnect" is presented.

Propose to have an option to select Web Notification, Email Notification and logged details of such actions (Agent Uninstalled!).

![notification](https://github.com/user-attachments/assets/c6d00935-d8ed-41f4-aac6-5408b487a2da)

~b10

**Comment by si458** [comment]
the problem you have with this is,
IF you uninstall the application from say the control panel
how do you know it was uninstalled?
you dont! all you would see is the websocket disconnected and never reconnected because the service got stopped and then removed
we would have to edit the meshagent to implement a function to watch and call if you tried to uninstall it,
we could already do this from the web ui because you selected uninstall, so we could change the notification to say agent uninstalled
but no way if the end user removed the application

**Comment by bayden10** [comment]
What I understand is if you use another installer package you could use a predefined Uninstall notification before the actual uninstall at the client side (Installed Apps/Uninstall). 

When the application is installed, you could register a command line to execute when the user clicks the Uninstall button. You can set that command line to do anything you want. For example, you can set it to Uninstall­String = "meshagent.exe" ⟨identification⟩ where ⟨identification⟩ is something that the agent can use to know what agent device is being uninstalled, so that it can launch the real uninstaller and update the registered meshcentral database etc with notifications. 

I know SplashTop has this ability to send a notification when the agent (streamer) is uninstalled and then also removes it from the list of devices (computer).

**Reply by si458** [reply]
let me double check this when i get chance as i know we have an `-uninstall` and `-fulluninstall` option from the cli
so we might be able to add something in to say do this when u start uninstalling, so we can tell meshcentral we are uninstalling

**Reply by si458** [reply]
ok so just checked source code

when u install meshagent it sets the uninstall string to `meshagent.exe -funinstall`
and in turn when you click uninstall it calls the exe with this option as a seperate process.
and it appears when we see `funinstall` we call the local uninstall commands embedded in the meshagent,
all this then simply does is, stop the agent service using windows service, systemctl, etc,
remove registry entries, remove files, then done

it does NOTHING ELSE, so meshcentral would have no way of knowing you uninstalled it!

we would have to physically modify the meshagent so it would connect to your meshcentral,
do all its security checks etc then push a command saying `yo i uninstalled`,
and the problem with that is it would create a second agent in your web ui!
OR
we would have to modify the meshagent for the 2nd process to talk to the main service and send it a command to say `yo im uninstalling`, then the meshagent would go `oh ok ill let meshcentral know`, that way you wouldnt create 2 agents by mistake

but what we can do is actuallly create events for the uninstall agent to keep a track of when staff asked to uninstall the agent using the meshctrl or web ui but not the physical remote device at the moment

so its very complicated to achieve what you wanted,

**Comment by bayden10** [comment]
Understood.. thank you for looking into this feature request.

**Reply by si458** [reply]
if you are happy with the idea of the being alerted to a staff asking meshcentral to uninstall, then open a github issue using the enhancement request option, and i can add in the option for you as indeed it does seem a good idea!
as currently when a staff member says uninstall or types uninstallagent the is no record of this anywhere!

just doing the physical monitoring on the client side is quite tricky at the moment as the agent doesnt get updated very often
because we can do alot of things server side now, just not the install/uninstall things sadly

**Reply by bayden10** [reply]
Will do.. 👍

**Comment by bayden10** [comment]
Opened #6618

**Reply by si458** [reply]
Great thank u!
I will have a look when I get chance and we may need to add it bit by bit but we will get there :)
